### PR TITLE
feature: SmartDocuments attended flow with templates

### PR DESCRIPTION
**Pull request by hsiliev**
Add redirect endpoint that triggers upload of information object
Use path and query parameters to store all needed information

Solves: PZ-3733

### CODE REVIEW

File: src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt [hunk 1]
```diff
@@ -47,4 +41,12 @@ interface SmartDocumentsClient {
         @HeaderParam("Authorization") authenticationToken: String,
         @HeaderParam("Username") userName: String
     ): SmartDocumentsTemplatesResponse
+
+    @GET
+    @Path("smartdocuments/result/show")
+    @Consumes("text/html; charset=UTF-8")
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt:47]: Is it not enough to do:

```suggestion
    @Consumes(MediaType.TEXT_HTML)
```
?

Is UTF8 not the default in our endpoints anyway?

File: src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt [hunk 2]
```diff
@@ -47,4 +41,12 @@ interface SmartDocumentsClient {
...
+    fun downloadFile(
+        @QueryParam("id") smartDocumentsId: String,
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt:49]: `smartDocumentsId` sounds a bit weird to me? Is it not just a:

```suggestion
        @QueryParam("id") fileId: String,
```

?

File: src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt [hunk 3]
```diff
@@ -9,98 +9,95 @@ import jakarta.inject.Inject
...
-import net.atos.zac.documentcreation.model.DocumentCreationDataUnattended
 import net.atos.zac.policy.PolicyService
 import net.atos.zac.policy.PolicyService.assertPolicy
 import nl.lifely.zac.util.AllOpen
 import nl.lifely.zac.util.NoArgConstructor
+import java.util.UUID
 
 @Singleton
-@Path("documentcreation")
-@Consumes(MediaType.APPLICATION_JSON)
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt:36]: Why did you remove these from the class level? Shouldn't the other two methods also use these?

File: src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt [hunk 4]
```diff
@@ -9,98 +9,95 @@ import jakarta.inject.Inject
...
-                    "zaaktype '${zaak.zaaktype}'. Cannot create document."
+
+    @POST
+    @Path("/redirect/zaak/{zaakUuid}")
+    fun createDocumentRedirect(
+        @PathParam("zaakUuid") zaakUuid: UUID,
+        @QueryParam("templateGroupId") templateGroupId: String,
+        @QueryParam("templateId") templateId: String,
+        @QueryParam("userName") userName: String,
+        @FormParam("sdDocument") smartDocumentId: String,
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt:66]: Same here. I find the terms `sdDocument` and `smartDocumentId` a bit confusing..

File: src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt [hunk 5]
```diff
@@ -9,98 +9,95 @@ import jakarta.inject.Inject
...
-                ).let(documentCreationService::createDocumentAttended)
-                    .let { RestDocumentCreationAttendedResponse(it.redirectUrl, it.message) }
-            } ?: run {
-            throw InputValidationFailedException(
-                "No informatieobjecttype '${ConfiguratieService.INFORMATIEOBJECTTYPE_OMSCHRIJVING_BIJLAGE}' found for " +
-                    "zaaktype '${zaak.zaaktype}'. Cannot create document."
+
+    @POST
+    @Path("/redirect/zaak/{zaakUuid}")
+    fun createDocumentRedirect(
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt:61]: So this is the endpoint that SmartDocuments will call when the document has been created? Does it need to be called `/redirect/zaak/..` or is that up to us? I would maybe call it something like `/smartdocuments/callback` or something. Also good to add some doc indicating that this is the callback URL that SD will invoke?

File: src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt [hunk 6]
```diff
@@ -9,98 +9,95 @@ import jakarta.inject.Inject
...
-                            zaak = zaak,
-                            enkelvoudigInformatieObjectCreateLockRequest = it,
-                            taskId = unattendedData.taskId
-                        )
-                    }
-                    RestDocumentCreationUnattendedResponse(message = unattendedResponse.message)
-                }
+    @Path("/redirect/zaak/{zaakUuid}/taak/{taakId}")
+    @Suppress("LongParameterList")
+    fun createDocumentRedirect(
```
Comment (edgarvonk) [inline src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt:83]: Another function called `createDocumentRedirect`? I'm getting a bit confused now..

File: src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt (hunk 3 above)
Comment (edgarvonk) [reply src/main/kotlin/net/atos/zac/app/documentcreation/DocumentCreationRestService.kt:36]: Bizarre..  But then can you not add `@Consumes(MediaType.TEXT_PLAIN)` to those endpoints? Good to make clear which mediatype we consume.

File: src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt (hunk 1 above)
Comment (hsiliev) [reply src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt]: > By default if no Content-Type header is present in a part, "text/plain; charset=us-ascii" is used as fallback. This is the value defined by the MIME RFC.

Comment (edgarvonk) [reply src/main/kotlin/net/atos/client/smartdocuments/SmartDocumentsClient.kt]: Yes, I know I think. We had issues with this in multi-part forms I believe where UTF8 was not uses because of this..

File: src/main/kotlin/net/atos/zac/documentcreation/converter/DocumentCreationDataConverter.kt [hunk 7]
```diff
@@ -195,4 +205,33 @@ class DocumentCreationDataConverter @Inject constructor(
...
+        userName: String
+    ) = EnkelvoudigInformatieObjectCreateLockRequest().apply {
+        bronorganisatie = ConfiguratieService.BRON_ORGANISATIE
+        creatiedatum = LocalDate.now()
+        titel = smartDocumentsFile.fileName
+        auteur = userName
+        taal = ConfiguratieService.TAAL_NEDERLANDS
+        beschrijving = ConfiguratieService.OMSCHRIJVING_TAAK_DOCUMENT
+        status = StatusEnum.DEFINITIEF
+        vertrouwelijkheidaanduiding = VertrouwelijkheidaanduidingEnum.OPENBAAR
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/documentcreation/converter/DocumentCreationDataConverter.kt:224]: That's what we already had in the existing converter, so I just kept the logic.

File: src/main/kotlin/net/atos/zac/documentcreation/DocumentCreationService.kt [hunk 8]
```diff
@@ -93,32 +66,49 @@ class DocumentCreationService @Inject constructor(
...
-                templateGroup = ztcClientService.readZaaktype(creationDataUnattended.zaak.zaaktype).omschrijving
+                templateGroup = smartDocumentsTemplatesService.getTemplateGroupName(
+                    creationDataUnattended.templateGroupId
+                ),
+                template = smartDocumentsTemplatesService.getTemplateName(creationDataUnattended.templateId),
+            ),
+            variables = Variables(
+                outputFormats = listOf(OutputFormat(OUTPUT_FORMAT_DOCX)),
+                redirectMethod = REDIRECT_METHOD,
+                redirectUrl = configuratieService.documentCreationUrl(
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/documentcreation/DocumentCreationService.kt:77]: Controversial indeed :)

Changed the function name to `documentCreationCallbackUrl`. I think the new name reflects the real purpose much better.

File: src/main/kotlin/net/atos/zac/configuratie/ConfiguratieService.kt [hunk 9]
```diff
@@ -139,6 +141,31 @@ class ConfiguratieService @Inject constructor(
...
+        zaakUuid: UUID,
+        taakUuid: String?,
+        templateGroupId: String,
+        templateId: String,
+        userName: String
+    ): URI =
+        if (taakUuid != null) {
+            UriBuilder
+                .fromUri(contextUrl)
+                .path("$SMART_DOCUMENTS_REDIRECT_URL_BASE/taak/{taakId}")
```
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/configuratie/ConfiguratieService.kt]: Changed variable to `taskId`.

[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/configuratie/ConfiguratieService.kt]: And the endpoint to `task`. I think we agreed zaak was special

File: src/main/kotlin/net/atos/zac/documentcreation/converter/DocumentCreationDataConverter.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (hsiliev) [reply src/main/kotlin/net/atos/zac/documentcreation/converter/DocumentCreationDataConverter.kt:224]: @bas-info-nl @camielb Any opinion on the status. Should we change the status to something else?

[the remark this answers is not in the document]
Comment (infonl-marcel) [reply src/main/kotlin/net/atos/zac/documentcreation/converter/DocumentCreationDataConverter.kt:224]: Maybe even more a question to Dimpact / Karin ? Same for vertrouwelijkheidaanduiding. But without knowing SD I would have expected those properties coming from SD itself. When entering a document. On our ZAC side we also ask for those props.

**Review by bas-info-nl — APPROVED** [review]
